librad2/src/keys/storage.rs

```diff
@@ -0,0 +1,285 @@
+use std::fs::File;
+use std::io;
+use std::io::prelude::*;
+use std::path::PathBuf;
+
+use secstr::SecUtf8;
+use sodiumoxide::crypto::pwhash;
+use sodiumoxide::crypto::secretbox;
+use sodiumoxide::crypto::sign;
+
+use crate::keys::types::DeviceKey;
+use crate::paths::Paths;
+
+#[derive(Debug)]
+pub enum Error {
+    KeyExists,
+    NoSuchKey,
+    KeyGenFailed,
+    InvalidSalt,
+    InvalidNonce,
+    InvalidKey,
+    InvalidPassphrase,
+    IoError(io::Error),
+}
+
+impl From<io::Error> for Error {
+    fn from(err: io::Error) -> Self {
+        Error::IoError(err)
+    }
+}
+
+pub trait Storage {
+    fn put_device_key<F>(&mut self, k: &DeviceKey, pinentry: F) -> Result<(), Error>
+    where
+        F: FnOnce() -> SecUtf8;
+
+    fn get_device_key<F>(&self, pinentry: F) -> Result<DeviceKey, Error>
+    where
+        F: FnOnce() -> SecUtf8;
+}
+
+#[derive(Default)]
+pub struct MemoryStorage {
+    device_key: Option<(DeviceKey, pwhash::HashedPassword)>,
+}
+
+impl MemoryStorage {
+    pub fn new() -> Self {
+        MemoryStorage { device_key: None }
+    }
+}
+
+impl Storage for MemoryStorage {
+    fn put_device_key<F>(&mut self, k: &DeviceKey, pinentry: F) -> Result<(), Error>
+    where
+        F: FnOnce() -> SecUtf8,
+    {
+        match self.device_key {
+            Some(_) => Err(Error::KeyExists),
+            None => {
+                let pwhash = pwhash::pwhash(
+                    pinentry().unsecure().as_bytes(),
+                    pwhash::OPSLIMIT_INTERACTIVE,
+                    pwhash::MEMLIMIT_INTERACTIVE,
+                )
+                .unwrap();
+                self.device_key = Some((k.clone(), pwhash));
+                Ok(())
+            }
+        }
+    }
+
+    fn get_device_key<F>(&self, pinentry: F) -> Result<DeviceKey, Error>
+    where
+        F: FnOnce() -> SecUtf8,
+    {
+        match self.device_key {
```

Comment:
Could you [`map_or`](https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or) this?